src/Bicep.Core/Workspaces/SourceFileGrouping.cs

```diff
@@ -38,13 +46,17 @@ public class SourceFileGrouping : IArtifactFileLookup
 
     public ImmutableDictionary<Uri, ResultWithDiagnostic<ISourceFile>> FileResultByUri { get; }
 
-    public ImmutableDictionary<BicepSourceFile, ImmutableDictionary<IArtifactReferenceSyntax, Result<Uri, UriResolutionError>>> FileUriResultByArtifactReference { get; }
+    public ImmutableDictionary<BicepSourceFile, ImmutableDictionary<IArtifactReferenceSyntax, Result<Uri, UriResolutionError>>> FileUriResultByBicepSourceFileByArtifactReferenceSyntax { get; }
+
+    public ImmutableDictionary<BicepSourceFile, ImmutableDictionary<ArtifactReference, Result<Uri, UriResolutionError>>> FileUriResultByBicepSourceFileByArtifactReference { get; }
+
+    public ImmutableDictionary<BicepSourceFile, ProviderDescriptorBundle> ProvidersToRestoreByFileResult { get; }
 
     public ImmutableDictionary<ISourceFile, ImmutableHashSet<ISourceFile>> SourceFileParentLookup { get; }
 
-    public IEnumerable<ArtifactResolutionInfo> GetArtifactsToRestore(bool force = false)
+    public IEnumerable<ArtifactResolutionInfo> GetExplicitArtifactsToRestore(bool force = false)
```

Comment:
What's the differentiation between "explicit" & "implicit", and why is it important to differentiate?

Reply:
All the artifacts that are processed by this function have a syntax node associated to it, so I introduce the concept of an `explicit` because it maps to a syntax node in the bicep source file. This contrasts with `implicit` ones since no such mapping to an `IArtifactReferenceSyntax` exists. They are declared in `bicepconfig.json` to be appended to the local scope.